[DI] Allow explicit taskLauncher for torchscript interpreter

garroud opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/46362

By default, TorchScript execution is single threaded and uses the caller's thread pool. For the use case of distributed inference, we hope there is a way to customize the behavior where the  interpreter in torch script can be executed in other places. This diff allows an explicit taskLauncher for torchscript interpreter.

Test Plan: unit test is passed.

Differential Revision: D24114874

fbshipit-source-id: 1d7b003926c0d1f8facc53206efb960cff8897ac

Fixes #{issue number}

houseroad commented on test/cpp/jit/test_graph_executor.cpp:
```diff
@@ -29,5 +32,40 @@ TEST(GraphExecutorTest, Basic_CUDA) {
 
+TEST(GraphExecutorTest, runAsync_executor) {
+  /*
+  TODO: there are some problem with C++ parsing script program involving
```
Do we have an issue on GIthub to track this issue? If so, we could add it in the comment.
If not, we could open one for tracking purpose.

houseroad commented on test/cpp/jit/test_graph_executor.cpp:
```diff
@@ -29,5 +32,40 @@ TEST(GraphExecutorTest, Basic_CUDA) {
+  auto testModelFile = filePath.substr(0, filePath.find_last_of("/\\") + 1);
+  testModelFile.append("test_interpreter_async.pt");
+  auto module = load(testModelFile);
+  module.to(at::kCPU);
```
wondering why we need to move the module to CPU, isn't the default device is cpu?

garroud replied:
it seems to be redundant, will remove.

houseroad reviewed (approved):
Looks good.

suo reviewed (approved):
lgtm also. Is there a followup diff where we show how we expect to be using this customization point? Right now the highest-level place where we are addign hooks is `GraphExecutor`, which is not particularly user-facing (users generally interact with stuff in `jit/api/`, like `Module` or `Function`.

Commit:
[DI] Allow explicit taskLauncher for torchscript interpreter (#46865)

Summary:
By default, TorchScript execution is single threaded and uses the caller's thread pool. For the use case of distributed inference, we hope there is a way to customize the behavior where the  interpreter in torch script can be executed in other places. This diff allows an explicit taskLauncher for torchscript interpreter.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/46865

Test Plan:
unit test is passed.

fbshipit-source-id: 1d7b003926c0d1f8facc53206efb960cff8897ac

Fixes #{issue number}

Reviewed By: houseroad

Differential Revision: D24616102

Pulled By: houseroad

fbshipit-source-id: 5e958fa8df9e8a3e3fff0580ccd7fcb318120a3d
